### ISSUE TITLE
MySQL connector update() function raises an exception without proceeding to the match block

### ISSUE THREAD

**Issue by NipunaRanasinghe** [ballerina-platform/ballerina-lang#8873]
**Description:**

I have been using mysql client endpoint update() function and when updating some records, the update() action seems to be throwing exceptions instead of letting user handle the error inside the match block.

**Steps to reproduce:**

Please refer the following sample code.

```ballerina
function upsertToPatchTable(json[] patches) returns boolean {
    log:printDebug(string `Preparing Upsert query for {{lengthof patches}} patches... `);
    string queryValues;

    string[] upsertedPatches = [];
    transaction with retries = 3, oncommit = onCommitJira, onabort = onAbortJira {
        int i = 0;
        upsertedPatches = [];
        foreach patch in patches {
            sql:Parameter jiraProjectKey = { sqlType: sql:TYPE_VARCHAR, value: patch["jira_project_key"].toString() };
            sql:Parameter jiraIssueKey = { sqlType: sql:TYPE_VARCHAR, value: patch["jira_issue_key"].toString() };
            sql:Parameter patchId = { sqlType: sql:TYPE_VARCHAR, value: patch["patch_id"].toString() };
            sql:Parameter date = { sqlType: sql:TYPE_VARCHAR, value: patch["date"].toString() };
            sql:Parameter uploader = { sqlType: sql:TYPE_VARCHAR, value: patch["uploader"].toString() };

            //converts the json array of related products into an comma separated single string
            string productsString = patch["products"].toString().replace(LEFT_SQUARE_BRACKET, EMPTY_STRING).
            replace(RIGHT_SQUARE_BRACKET, EMPTY_STRING).replace(DOUBLE_QUOTATION_MARKS, EMPTY_STRING);

            sql:Parameter products = { sqlType: sql:TYPE_VARCHAR, value: productsString };
            log:printDebug("update starting for the patch: " + patch["patch_id"].toString());

           //exceptions is raised here (only for some records) 
            var results = mysqlEP->update(QUERY_TEMPLATE_UPSERT_PATCHES, jiraProjectKey, jiraIssueKey, patchId, date
                , uploader, products);
            match results {
                int c => {
                    log:printDebug(string `Upserting patch {{i}} -> Return value: {{c}}`);
                    if (c < 0) {
                        log:printError("Unable to Upsert to Patch table");
                        abort;
                    } else {
                        log:printDebug("Successfully Upserted to patch table");
                        upsertedPatches[i] = patch["patch_id"].toString();
                        i++;
                    }
                }
                error e => {
                    log:printDebug("db call failed: " + e.message);
                }
            }
        }
    } onretry {
        log:printWarn("Retrying transaction to upsert to patch table ");
    }

    // Check all patches were inserted and also transaction is commited
    return lengthof upsertedPatches == lengthof patches;
}
```
**Affected Versions:**
ballerina v0.970.0

**Suggested Labels (optional):**
bug, component/stdlib

**Suggested Assignees (optional):**

**Comment by Manuri** [comment]
Following is a simpler program to reproduce the issue.

```ballerina
import ballerina/mysql;
import ballerina/io;

endpoint mysql:Client testDB {
    host: "localhost",
    port: 3306,
    name: "testdb",
    username: "root",
    password: "123",
    poolOptions: { maximumPoolSize: 5 },
    dbOptions: { useSSL: false }
};

function main(string... args) {
    transaction with retries = 3,  oncommit = onCommitFunc, onabort = onAbortFunc {
        try {
            var ret = testDB->update("insert into test values(?, ?)", 2, 2);

            match ret {
                int => io:println("int ret");
                error => io:println("error ret");
            }
        } catch (error e) {
            io:println(e.message);
        }
    }
}

function onCommitFunc(string id) {
    io:println("committed");
}

function onAbortFunc(string id) {
    io:println("aborted");
}
```

Output:
```bash
ballerina: started HTTP/WS endpoint 172.17.0.1:33369
2018-08-21 14:00:33,785 INFO  [ballerina.transactions] - Created transaction: f1cbdd2f-bb80-4d6e-8b32-04fa4bbeb474 
TransactionError
2018-08-21 14:00:33,810 INFO  [ballerina.transactions] - Running 2-phase commit for transaction: f1cbdd2f-bb80-4d6e-8b32-04fa4bbeb474:1 
committed
ballerina: stopped HTTP/WS endpoint 172.17.0.1:33369
```

**Comment by anupama-pathirage** [comment]
This is fixed with 0.990.0 release. 

Checked with the below code.

```ballerina
import ballerina/io;
import ballerina/h2;

h2:Client testDB = new ({
    path: "./local-transactions/",
    name: "Testdb",
    username: "test",
    password: "test",
    poolOptions: { maximumPoolSize: 5 }
});

public function main() {
    var ret = testDB->update("CREATE TABLE CUSTOMER (ID INTEGER, NAME
                              VARCHAR(30))");
    handleUpdate(ret, "Create CUSTOMER table");    ret = testDB->update("CREATE TABLE SALARY (ID INTEGER, MON_SALARY FLOAT)");
    handleUpdate(ret, "Create SALARY table");
    transaction with retries = 4 {
        var count = testDB->update("INSERT INTO CUSTOMER(ID,NAME)
                                     VALUES (1, 'Anne')");
        count = testDB->update("INSERT INTO SALARYd (ID, MON_SALARY)
                                 VALUES (1, 2500)");
        if (count is int) {
            io:println("Inserted count: " + count);
            if (count == 0) {
                abort;
            }
        } else {
            io:println("ERROR OCCURED:");
            io:println(count);
        }
    } onretry {
        io:println("Retrying transaction");
    } committed {
        io:println("Transaction committed");
    } aborted {
        io:println("Transaction aborted");
    }
    ret = testDB->update("DROP TABLE CUSTOMER");
    handleUpdate(ret, "Drop table CUSTOMER");    ret = testDB->update("DROP TABLE SALARY");
    handleUpdate(ret, "Drop table SALARY");
    testDB.stop();
}
function handleUpdate(int|error returned, string message) {
    if (returned is int) {
        io:println(message + " status: " + returned);
    } else {
        io:println(message + " failed: " + returned.reason());
    }
}

```
